### PR TITLE
Add Experts with distributed exec

### PR DESCRIPTION
**Pull request by pprajapatiTT**
### Ticket
None

### Problem description
Change expert.py to execute deepseek experts on TT hardware. This code creates `n_routed_experts//num_of_devices` groups and executes each group on a device.

### What's changed
Change expert.py
Add test_moe_experts.py to test the 'Expert' module

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/demos/deepseek_v3/tt/expert.py [hunk 1]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
         dim = hf_config.hidden_size
         hidden_dim = hf_config.moe_intermediate_size
         return dim, hidden_dim
+
+    @classmethod
+    def convert_weights_moe(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py:39]: You don't seem to be dequantizing the weights here

File: models/demos/deepseek_v3/tt/expert.py [hunk 2]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        # breakpoint()  # Debugging point to inspect the state_dict structure
+        num_experts_per_device = hf_config.n_routed_experts // mesh_device.get_num_devices()
+        num_groups = hf_config.n_routed_experts // num_experts_per_device
+
+        w1_per_device_state_dict_group = []
+        w2_per_device_state_dict_group = []
+        w3_per_device_state_dict_group = []
+
+        # Group experts by device
+        for device_idx in range(mesh_device.get_num_devices()):
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: can't you perhaps loop over the weight types just like in `MLP1D` to reduce code duplication?

File: models/demos/deepseek_v3/tt/expert.py [hunk 3]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        cls,
+        hf_config: PretrainedConfig,
+        state_dict: dict[str, torch.Tensor],
+        output_path: Path,
+        mesh_device: ttnn.Device,
+    ) -> WeightConfig:
+        # assert cls.is_device_supported(mesh_device)
+
+        # breakpoint()  # Debugging point to inspect the state_dict structure
+        num_experts_per_device = hf_config.n_routed_experts // mesh_device.get_num_devices()
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: Please use `even_int_div` from config_helpers, it's safer

File: models/demos/deepseek_v3/tt/expert.py [hunk 4]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        num_groups = hf_config.n_routed_experts // num_experts_per_device
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: Same here

File: models/demos/deepseek_v3/tt/expert.py [hunk 5]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+                memory_config=ttnn.L1_MEMORY_CONFIG,
+                input_tensor_a_activations=[ttnn.UnaryOpType.SILU],
+            ),
+            "input_memory_config": input_memory_config,  # For asserting the input to the MLP
+            "output_memory_config": output_memory_config,  # For asserting the output of the MLP
+            "num_experts_per_device": num_experts_per_device,
+        }
+
+    @classmethod
+    def forward_prefill(cls, x: ttnn.Tensor, cfg: RunPrefillConfig) -> ttnn.Tensor:
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: Is this in any way different to the MLP1D forward_prefill logic, apart from the config names?

File: models/demos/deepseek_v3/tt/expert.py [hunk 6]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        # De-chunk the output if the input was chunked
+        _, num_chunks, _, output_dim = output.shape
+        if num_chunks > 1:
+            output = ttnn.reshape(output, [1, 1, -1, output_dim])
+
+        assert output.memory_config() == cfg["output_memory_config"]
+        return output
+
+    @classmethod
+    def forward_decode(cls, x: ttnn.Tensor, cfg: RunDecodeConfig) -> ttnn.Tensor:
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: Is this in any way different to the MLP1D forward_decode logic, apart from the config names?

File: models/demos/deepseek_v3/tt/expert.py (hunk 1 above)
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py:39]: I wonder if instead of reusing the weight saving logic here we could concatenate the per-device-weights and/or modify the MLP1D logic to support doing that with lower-level method overwrites?

**Review by barci2 — CHANGES_REQUESTED** [review]

File: models/demos/deepseek_v3/tt/expert.py [hunk 7]
```diff
@@ -15,3 +39,344 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        assert x.memory_config() == cfg["input_memory_config"]
+        breakpoint()
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/expert.py]: Please remove breakpoints before merging!

File: models/demos/deepseek_v3/tt/expert.py (hunk 6 above)
Comment (yieldthought) [reply models/demos/deepseek_v3/tt/expert.py]: Reshaping based on num experts?

**Comment by pprajapatiTT** [comment]
@barci2 : thanks for the comments. This is draft/WIP PR. Feel free to revise the comments when we mark the PR ready for review.

File: models/demos/deepseek_v3/tt/expert.py (hunk 1 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py:39]: Yes, I added a TODO. We'll revise it in separate patch.

File: models/demos/deepseek_v3/tt/expert.py (hunk 2 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: We need to create groups of experts so cann't reuse exact code as MLP1D.

File: models/demos/deepseek_v3/tt/expert.py (hunk 5 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: Group of experts are running on a device, unlike MLP1D.

File: models/demos/deepseek_v3/tt/expert.py (hunk 6 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: > Reshaping based on num experts?
Sorry i didn't get main context. What is your recommendation?

Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: > Is this in any way different to the MLP1D forward_decode logic, apart from the config names?

Group of experts are running on a device, unlike MLP1D.

File: models/demos/deepseek_v3/tt/expert.py (hunk 1 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py:39]: That would complicate MLP1D, I like to keep MLP1D simple so many other modules can use it in simple way. Parametrizing it with many options can cause issues later in maintenance of the code.

File: models/demos/deepseek_v3/tt/expert.py (hunk 7 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: yup. removed

**Comment by pprajapatiTT** [comment]
Marked this PR as ready to review. I addressed previous comments. Let me know if you have more!

**Comment by pprajapatiTT** [comment]
Also note that, dram sharded matmul not supported in this case as we have 3D weights

File: models/demos/deepseek_v3/tt/expert.py (hunk 2 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/expert.py]: yes but you can have a per-weight-type dict of dicts for that...

File: models/demos/deepseek_v3/tt/expert.py [hunk 8]
```diff
@@ -15,3 +34,192 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        return {
+            "w1_experts": LinearConfig(
+                input_tensor_b=FromWeightConfig(MeshDeviceStub(mesh_device.shape)),
+                memory_config=mem_config,
+                compute_kernel_config=COMPUTE_KERNEL_CONFIG_LOFI,
+            ),
+            "w2_experts": LinearConfig(
+                input_tensor_b=FromWeightConfig(MeshDeviceStub(mesh_device.shape)),
+                memory_config=mem_config,
+                compute_kernel_config=COMPUTE_KERNEL_CONFIG_SDPA,
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/expert.py]: Using the SDPA compute kernel here is surely a mistake?

**Review by yieldthought — CHANGES_REQUESTED** [review]

File: models/demos/deepseek_v3/reference/modeling_deepseek.py [hunk 9]
```diff
@@ -407,6 +407,31 @@ def forward(self, hidden_states):
         return topk_idx, topk_weight
 
 
+class DeepseekV3MoE_Experts(nn.Module):
```
Comment (barci2) [inline models/demos/deepseek_v3/reference/modeling_deepseek.py]: let's maybe keep this the same as the HF version, and just move this class or forward function to the test itself

File: models/demos/deepseek_v3/tests/test_moe_experts.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    hf_config.n_routed_experts = 64  # Set number of experts for testing
+    return DeepseekV3MoE_Experts(hf_config)
+
+
+@pytest.mark.parametrize(
+    "mode,seq_len",
+    [
+        ("decode", 128),
+        ("prefill", 256),
+        # ("prefill", 2048),
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_moe_experts.py]: are we not testing the sequence length 2048?

File: models/demos/deepseek_v3/tests/test_moe_experts.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    ],
+)
+def test_forward_pass(
+    mode,
+    seq_len,
+    reference_model,
+    hf_config_single_layer,
+    temp_dir,
+    galaxy_or_t3k_mesh,
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_moe_experts.py]: please add an assertion that the test is only run on the full glx (is it necessary)?

File: models/demos/deepseek_v3/tests/test_moe_experts.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    ), f"Output memory config mismatch: expected {expected_output_memory_config}, got {actual_output_memory_config}"
+
+    # output shape per device  = [1, experts_per_device, seq_len, hidden_size]
+    # There are 32 groups of unique experts output in case of TG
+    # We first concate rows and then columns to get the final output
+    # Convert output back to torch
+    tt_output_torch = ttnn.to_torch(
+        tt_output,
+        mesh_composer=ttnn.ConcatMesh2dToTensor(
+            galaxy_or_t3k_mesh, dims=(0, 1), mesh_shape=tuple(galaxy_or_t3k_mesh.shape)
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_moe_experts.py:130]: maybe change it to (-2, -1) in case there are 4 output dims?

File: models/demos/deepseek_v3/reference/modeling_deepseek.py (hunk 9 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/reference/modeling_deepseek.py]: It's better to keep this here only. I am not changing original model creation interface so it would not affect anybody.

File: models/demos/deepseek_v3/tt/expert.py (hunk 2 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: I don't see any advantage in using dict of dicts here.

File: models/demos/deepseek_v3/tt/expert.py (hunk 8 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: yes, This is fixed now.

File: models/demos/deepseek_v3/tests/test_moe_experts.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC.
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_moe_experts.py:1]: Once [this pr](https://github.com/tenstorrent/tt-metal/pull/25509) is merged, could you add real IO tests to MoE and expert tests?

File: models/demos/deepseek_v3/tt/expert.py (hunk 2 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/expert.py]: Well the advantage is that you have 3x less code

File: models/demos/deepseek_v3/reference/modeling_deepseek.py (hunk 9 above)
Comment (barci2) [reply models/demos/deepseek_v3/reference/modeling_deepseek.py]: @yieldthought thoughts?

File: models/demos/deepseek_v3/tt/expert.py [hunk 14]
```diff
@@ -15,3 +29,192 @@ def _get_model_dims_from_cfg(cls, hf_config: PretrainedConfig) -> tuple[int, int
...
+        # Down projection
+        output = ttnn.linear(activated, **cfg["w2_experts"])
+        ttnn.deallocate(activated)
+
+        assert output.memory_config() == cfg["output_memory_config"]
+        return output
+
+    @classmethod
+    def forward_prefill(cls, x: ttnn.Tensor, cfg: RunPrefillConfig) -> ttnn.Tensor:
+        return cls.forward_decode(x, cfg)
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/expert.py]: maybe have a separate function that both prefill and decode calls? This gives an impression that prefill is somehow dependent or an extension of decode

File: models/demos/deepseek_v3/tests/test_moe_experts.py (hunk 12 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tests/test_moe_experts.py:130]: (0, 1) is correct only. Output tensors is already 4D. (0, 1) are the dims of output tensors which we want to concatenate.
In minus notation it will be (-3, -2) which doesn't safe us if its 3D or 4D.

File: models/demos/deepseek_v3/tests/test_moe_experts.py (hunk 13 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_moe_experts.py:1]: yes. I added a TODO in code.

File: models/demos/deepseek_v3/tt/expert.py (hunk 14 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/expert.py]: done

File: models/demos/deepseek_v3/tests/test_moe_experts.py (hunk 10 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_moe_experts.py]: Will add 2048 and remove 256

File: models/demos/deepseek_v3/tests/test_moe_experts.py (hunk 11 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_moe_experts.py]: Not needed. This works for both galaxy and t3k.

**Review by kpaigwar — APPROVED** [review]
looks good to me!

**Review by yieldthought — APPROVED** [review]
🙌

**Review by barci2 — APPROVED** [review]
lgtm
